sunscreen_zkp_backend/src/bulletproofs.rs

```diff
@@ -1,77 +1,375 @@
+                }
+                Operation::Sub => {
+                    let (left, right) = query.get_binary_operands(idx)?;
+
+                    let left = self.nodes[left.index()]
+                        .as_ref()
+                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(left)))
+                        .clone();
+
+                    let right = self.nodes[right.index()]
+                        .as_ref()
+                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(right)))
+                        .clone();
+
+                    self.nodes[idx.index()] = Some(left - right);
+                }
+                Operation::Neg => {
+                    let left = query.get_unary_operand(idx)?;
+
+                    let left = self.nodes[left.index()]
+                        .as_ref()
+                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(left)))
+                        .clone();
+
+                    self.nodes[idx.index()] = Some(-left);
+                }
+                Operation::Mul => {
+                    let (left, right) = query.get_binary_operands(idx)?;
 
-        let mut prover = Prover::new(&pc_gens, transcript);
+                    let left = self.nodes[left.index()]
+                        .as_ref()
+                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(left)))
+                        .clone();
 
-        let inputs = vec![
-            prover.allocate(Some(x)).unwrap(),
-            prover.allocate(Some(y)).unwrap(),
-        ];
+                    let right = self.nodes[right.index()]
+                        .as_ref()
+                        .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(right)))
+                        .clone();
 
-        let outputs = vec![prover.allocate(Some(o)).unwrap()];
+                    let (_, _, o) = cs.multiply(left, right);
 
-        Self::gadget(&mut prover, inputs, outputs);
+                    self.nodes[idx.index()] = Some(o.into());
+                }
+                Operation::Constraint(x) => {
+                    let operands = query.get_unordered_operands(idx)?;
 
-        Self(prover.prove(&bp_gens).unwrap())
+                    let x = try_uint_to_scalar(&x)?;
+
+                    for o in operands {
+                        let o = self.nodes[o.index()]
+                            .as_ref()
+                            .unwrap_or_else(|| panic!("{}", dependency_not_found_msg(o)))
+                            .clone();
+
+                        cs.constrain(o - x);
+                    }
+                }
+            }
+
+            Ok::<(), Error>(())
+        })?;
+
+        Ok(())
     }
+}
 
-    pub fn verify(&self) -> bool {
-        let (transcript, pc_gens, bp_gens) = Self::make_gens();
+impl ZkpProverBackend for BulletproofsR1CSCircuit {
+    fn prove(mut graph: ZkpBackendCompilationResult, inputs: &[BigInt]) -> Result<Box<Proof>> {
```

Comment:
Kinda gross that graph has to be passed as mut when in fact it won't be mutated.

This is because I didn't want to implement an immutable traversal function, so we're using the mutable one and just not returning any transforms so the graph doesn't mutate.

Reply:
Fixed.